src/meta/src/hummock/compaction/tier_compaction_picker.rs

```diff
@@ -0,0 +1,361 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use bytes::Bytes;
+use risingwave_hummock_sdk::key::{user_key, FullKey};
+use risingwave_hummock_sdk::key_range::KeyRange;
+use risingwave_hummock_sdk::HummockEpoch;
+use risingwave_pb::hummock::{Level, LevelType, SstableInfo};
+
+use super::SearchResult;
+use crate::hummock::level_handler::LevelHandler;
+
+const DEFAULT_MAX_COMPACTION_BYTES: u64 = 2 * 1024 * 1024 * 1024; // 2GB
+const DEFAULT_LEVEL0_MAX_FILE_NUMBER: usize = 16;
+const DEFAULT_LEVEL0_TRIGGER_NUMBER: usize = 4;
+
+#[derive(Debug)]
+pub struct TierCompactionPicker {
+    compact_task_id: u64,
+    max_compaction_bytes: u64,
+    level0_max_file_number: usize,
+    level0_trigger_number: usize,
+}
+
+#[derive(Default)]
+pub struct ConflictKeyRange {
+    key_range: Option<KeyRange>,
+}
+
+impl ConflictKeyRange {
+    fn check_table_conflict(&mut self, level_handler: &LevelHandler, table: &SstableInfo) -> bool {
+        let key_range = KeyRange::from(table.key_range.as_ref().unwrap());
+        if level_handler.is_pending_compact(&table.id) {
+            if let Some(range) = self.key_range.as_mut() {
+                range.full_key_extend(&key_range);
+            } else {
+                self.key_range = Some(key_range);
+            }
+            return true;
+        }
+        if let Some(range) = self.key_range.as_ref() {
+            // This file is conflict with previous files in L0.
+            if key_range.full_key_overlap(range) {
+                return true;
+            }
+        }
+        false
+    }
+}
+
+impl TierCompactionPicker {
+    pub fn new(compact_task_id: u64) -> TierCompactionPicker {
+        TierCompactionPicker {
+            compact_task_id,
+            // TODO: set it by cluster configure or vertical group configure.
+            max_compaction_bytes: DEFAULT_MAX_COMPACTION_BYTES,
+            level0_max_file_number: DEFAULT_LEVEL0_MAX_FILE_NUMBER,
+            level0_trigger_number: DEFAULT_LEVEL0_TRIGGER_NUMBER,
+        }
+    }
+
+    pub fn pick_compaction(
+        &self,
+        levels: Vec<Level>,
+        level_handlers: &mut [LevelHandler],
+    ) -> Option<SearchResult> {
+        let select_level = 0;
+        // TODO: After support dynamic-leveled-compaction, use base-level as target-level.
+        let target_level = 1;
+
+        let next_task_id = self.compact_task_id;
+        if levels[select_level].table_infos.is_empty() {
+            return None;
+        }
+        let (select_level_inputs, target_level_inputs) = self.select_input_files(
+            &levels[select_level],
+            &levels[target_level],
```

Comment:
`levels[0]` is sorted by epoch and `levels[1]` is sorted by key range, isn't it?

Reply:
I sort them when apply compact result